### ISSUE TITLE
Editor scrollbar covers class and function selector; console scrollbar covers console pane

### ISSUE THREAD

**Issue by mrclary** [spyder-ide/spyder#13118]
### Issue Report Checklist

* [ x] Searched the [issues page](https://github.com/spyder-ide/spyder/issues?q=is%3Aissue) for similar reports
* [ x] Read the relevant sections of the [Spyder Troubleshooting Guide](https://github.com/spyder-ide/spyder/wiki/Troubleshooting-Guide-and-FAQ) and followed its advice
* [ x] Reproduced the issue after updating with ``conda update spyder`` (or ``pip``, if not using Anaconda)
* [ x] Could not reproduce inside ``jupyter qtconsole`` (if console-related)
* [ x] Tried basic troubleshooting (if a bug/error)
    * [ x] Restarted Spyder
    * [ x] Reset preferences with ``spyder --reset``
    * [ x] Reinstalled the latest version of [Anaconda](https://www.anaconda.com/download/)
    * [ x] Tried the other applicable steps from the Troubleshooting Guide
* [ x] Completed the **Problem Description**, **Steps to Reproduce** and **Version** sections below

## Problem Description
On MacOS, the Editor scrollbar covers the dropdown arrow in the class and function selector

<img width="585" alt="Screen Shot 2020-06-26 at 4 41 58 PM" src="https://user-images.githubusercontent.com/9618975/85908914-17e02b80-b7cc-11ea-8c3d-b510f3ad2b1e.png">

It seems that the scrollbar is too tall and does not line up with the scrollflag area (should be shortened to reveal the selector arrow)

Also, the console scrollbar covers the console pane, obscuring text on long lines

<img width="814" alt="Screen Shot 2020-06-26 at 4 36 43 PM" src="https://user-images.githubusercontent.com/9618975/85908964-6392d500-b7cc-11ea-8894-9f45cd31ba4b.png">

Here, the "wn" is obscured.

### What is the expected output? What do you see instead?

Unobscured elements

I figured @steff456 did such a good job fixing #12637 that she'd make quick work of these issues!

**Comment by juanis2112** [comment]
@steff456 can you take a look at this?

**Comment by mrclary** [comment]
This appears to be caused by `qdarkstyle`.
The following minimum working example illustrates the issue.
Simply comment/uncomment `self.setStyleSheet(str(stylesheet))` to contrast the behavior.

```python
import sys
from PyQt5.QtWidgets import QApplication, QMainWindow, QTextEdit
import qdarkstyle


class TextEditor(QMainWindow):
    def __init__(self):
        super().__init__()

        self.setWindowTitle("Simple Text Editor")
        self.setGeometry(100, 100, 400, 400)

        # Text area
        self.text_edit = QTextEdit()
        self.text_edit.setText(
            "The quick brown fox jumped over the lazy dog. The quick brown fox jumped over the lazy dog. The quick brown fox jumped over the lazy dog. The quick brown fox jumped over the lazy dog. The quick brown fox jumped over the lazy dog. The quick brown fox jumped over the lazy dog. The quick brown fox jumped over the lazy dog. The quick brown fox jumped over the lazy dog. The quick brown fox jumped over the lazy dog. The quick brown fox jumped over the lazy dog. The quick brown fox jumped over the lazy dog. The quick brown fox jumped over the lazy dog. The quick brown fox jumped over the lazy dog. The quick brown fox jumped over the lazy dog. The quick brown fox jumped over the lazy dog. The quick brown fox jumped over the lazy dog. The quick brown fox jumped over the lazy dog. The quick brown fox jumped over the lazy dog. The quick brown fox jumped over the lazy dog. The quick brown fox jumped over the lazy dog. The quick brown fox jumped over the lazy dog. The quick brown fox jumped over the lazy dog. The quick brown fox jumped over the lazy dog. The quick brown fox jumped over the lazy dog. The quick brown fox jumped over the lazy dog. The quick brown fox jumped over the lazy dog. The quick brown fox jumped over the lazy dog. The quick brown fox jumped over the lazy dog. The quick brown fox jumped over the lazy dog. The quick brown fox jumped over the lazy dog. The quick brown fox jumped over the lazy dog. The quick brown fox jumped over the lazy dog. The quick brown fox jumped over the lazy dog. The quick brown fox jumped over the lazy dog. The quick brown fox jumped over the lazy dog. The quick brown fox jumped over the lazy dog. The quick brown fox jumped over the lazy dog. The quick brown fox jumped over the lazy dog. The quick brown fox jumped over the lazy dog. The quick brown fox jumped over the lazy dog. The quick brown fox jumped over the lazy dog. The quick brown fox jumped over the lazy dog. The quick brown fox jumped over the lazy dog. The quick brown fox jumped over the lazy dog. The quick brown fox jumped over the lazy dog. The quick brown fox jumped over the lazy dog. The quick brown fox jumped over the lazy dog."
        )
        self.setCentralWidget(self.text_edit)

        stylesheet = qdarkstyle.load_stylesheet()
        # Setting the stylesheet from qdarkstyle causes the scroll bar issue.
        # Comment the following line to change behavior
        self.setStyleSheet(str(stylesheet))


if __name__ == "__main__":
    app = QApplication(sys.argv)
    editor = TextEditor()
    editor.show()
    sys.exit(app.exec_())

```

**Comment by mrclary** [comment]
<img width="410" height="433" alt="Image" src="https://github.com/user-attachments/assets/a5a0a1b9-79fc-49c9-860b-c58767ef443e" />

<img width="404" height="432" alt="Image" src="https://github.com/user-attachments/assets/750791d2-88b0-4e6a-abec-7ebf94ee6ef2" />

<img width="408" height="435" alt="Image" src="https://github.com/user-attachments/assets/45799b43-06e7-4d20-a004-95011c0f36c9" />

<img width="405" height="432" alt="Image" src="https://github.com/user-attachments/assets/563b4fd4-d1c6-46fa-af0e-a078773eb894" />

**Comment by ccordoba12** [comment]
> This appears to be caused by qdarkstyle.

Great finding @mrclary!

Maybe qdarkstyle is forcing scrollbars to always be visible, so we need to fix that.

**Comment by mrclary** [comment]
What I've learned so far...

### tl;dr

This issue is not a `qdarkstyle` or `Qt` bug, as far as I can tell, but is an inherent result of how the style sheets work. I don't think there is a way to revert to the system scroll bars and maintain the full `qdarkstyle` style. However, we can set macOS's "Show scroll bars" preference to "Always" for just Spyder by adding a command to the `post-link.sh` script in Spyder's feedstock. This is what I recommend.

### Qdarkstyle

I've found that removing the following items from the style sheet results in reverting to the system scroll bars.
* `stylesheet._child_rules["*"].pop("border")`
* `stylesheet._child_rules["*"].pop("border-style")`
* `stylesheet._child_rules["QWidget"].pop("background-color")  # or set ""/None`
* `stylesheet._child_rules["QWidget"].pop("border")`
* `stylesheet._child_rules["QScrollBar:vertical"].clear()`
* `stylesheet._child_rules["QScrollBar:horizontal"].clear()`

Unfortunately, removing the global (`"*"`) and `"QWidget"` style rules breaks a lot of other UI elements.
I don't believe there is a way to have `QScrollBar` _not_ inherit from the global or `"QWidget"` rules. Once these are set, with any value, even null or empty ones, the system scroll bars are abandoned, so it is not even possible to overwrite all `QScrollBar` styles with null or empty values. This results in null or empty style rules, but does not revert to system scroll bars.

I also tried to "imitate" system scroll bar behavior by making scroll bar elements semi-transparent. This was successful insofar as being able to adjust visual aspects of the scroll bars to not obstruct scroll area content. However, I think the UI would be a very poor one and cause too much confusion for users since the scroll bars then become more difficult to discern. It also does not solve the issue of overlapping border images (scroll arrows).

### macOS Preferences

macOS provides a System Preference to show scroll bars "Always", which resolves all of the issues. However, this affects all applications system-wide, so is not a desirable solution. Fortunately, this preference can be set at a per-app level with the command:
```
defaults write <app-domain> AppleShowScrollBars Always
```
The drawback is that will only apply to applications with an application domain specified in the `info.plist`. This means that any Spyder application installed via `conda` (and `menuinst`), i.e. from `conda-forge` or our installers can have this preference applied. Unfortunately any other installation (`pip`, or developer), cannot have this preference applied.

In order to implement this, we could add the above command to the `post-link.sh` script in our feedstock.

**Comment by ccordoba12** [comment]
> In order to implement this, we could add the above command to the post-link.sh script in our feedstock.

Sounds reasonable to me. Thanks for taking the time to look deeply into this @mrclary!

As part of this, you could also remove the big scrollflag width in the editor for conda installations, which was added to address the problem with auto-hiding scrollbars on Mac.

**Comment by mrclary** [comment]
I'll create a PR on the `rc` branch of the feedstock, but we can wait to merge until after 6.1.4 is released.
